keps/sig-etcd/4743-kuberernetes-etcd-interface/README.md

```diff
@@ -0,0 +1,349 @@
+This interface will be used for etcd robustness tests and adopted by Kubernetes
+to validate the correctness of its etcd contract usage.
+
+## User Stories
+
+To better understand the importance of code location let’s visit the following use cases:
+
+### Kubernetes Backporting a etcd client patch version
+
+**The Journey:** Kubernetes regularly updates to newer etcd versions to leverage
+bug fixes, or security patches. However, ensuring compatibility between the 
+codified etcd-Kubernetes interface and etcd client is essential.
+
+**Considerations:**
+
+*   Even minor etcd client updates might inadvertently introduce changes that 
+    break the interface's assumptions or functionality.
+*   Kubernetes needs to choose a compatible interface version before upgrading 
+    the etcd client.
+*   Tight coupling to the etcd client could necessitate backporting the 
+    interface to older etcd branches, a complex and time-consuming process.
+
+
+### Making Changes/Patching Bugs in the Interface
+
+**The Journey:** Despite careful design, the complex etcd-Kubernetes contract 
+might reveal bugs or require adjustments.
+
+**Considerations:**
+
+*   Changes and bug fixes need to be implemented and released with minimal
+    disruption to both Kubernetes and etcd users.
+*   Tightly coupled interfaces (Options 1 and 2) might require a full etcd
+    client release for bug fixes, slowing down the process.
+
+### Kubernetes Leveraging New etcd Functionality
+
+**The Journey:** Kubernetes wants to expand its use of the etcd API beyond the
+current interface's scope.
+
+**Considerations:** The interface codifies a minimal subset of the etcd API
+currently used by Kubernetes, and new features will initially be outside its
+scope. Balancing new feature adoption with interface stability is crucial.
+
+**Mitigation:**
+
+*   Allow Kubernetes to directly use new etcd client features during alpha/beta
+    stages, bypassing the interface temporarily.
+*   Extending etcd robustness test to cover the new functionality before
+    formalizing them in interface.
+*   Once a feature is mature and stable, extend the interface, ensuring backward
+    compatibility for existing Kubernetes versions.
+
+## Code location
+
+We propose locating the interface in a `kubernetes` subdirectory under
+https://github.com/etcd-io/etcd/tree/main/client/v3. 
+This approach allows for seamless integration with the etcd client while 
+maintaining a dedicated space for the interface code.
+
+Interfaced will be part of the etcd client package and it's release will be
+combined with etcd release. For immediate Kubernetes use etcd will backport the 
+client to `release-3.5` branch and introduce it in next etcd patch release for 
+Kubernetes to consume.
+
+Alternative code locations are discussed at the end of the document.
+
+## The Interface
+
+To ensure smoother transition I propose the adoption of the etcd-Kubernetes interface to be done in two stages:
```

Comment:
Avoid using “I”; the KEP might ultimately end up as the work of multiple contributors.

Reply:
Oh, sorry it was left from the original document